### ISSUE TITLE
RFC: Moving to git-flow type development

### ISSUE THREAD

**Issue by roryk** [bcbio/bcbio-nextgen#2881]
Hey everyone,

bioconda has a nice bot which automatically updates the bioconda packages when there is a new release, so there is more reason than ever to switch to a more git-flow type development, as we can push patches to master, cut patch-releases and they will automatically get built into the bioconda package.

This will require a few changes, right now our development flow looks like this:

1. master should be bug free. new features and patches both get merged directly into master once they are shown to work
2. new features get forked from master, worked on their own branch and pushed back to master when they are complete
3. bug fixes get pushed directly into master
4. new releases happen once we've stabilized new features or there are major bug fixes that need to be in every version of bcbio
5. we increment version numbers as 1.xx.

This flow has very little overhead, but has some disadvantages. Namely, if we push new features or changes to master, users cannot get bug fixes without also pulling in the new features. It also makes 'master' a little less stable than it could be, since new features appear in master which is destabilizing.

The git-flow workflow looks like this:

1. A develop branch is created from master
2. A release branch is created from develop
3. Feature branches are created from develop
4. When a feature is complete it is merged into the develop branch
5. When the release branch is done it is merged into develop and master
6. If an issue in master is detected a hotfix branch is created from master
7. Once the hotfix is complete it is merged to both develop and master

so there is not that much extra overhead, mostly just to remember to implement new features from develop instead of on master. We can still do bugfixes to master like we have been doing. This will let us keep master more stable than it is currently and let users stick with it for longer while we work on other, non bugfix improvements. It also lets us use the new bioconda bot and keep our packages on bioconda up to date, bug-free and stable.

I think this is a good switch, it doesn't add very much process and has some useful benefits to it. There is a git-flow tool that makes these processes a little simpler to do: https://danielkummer.github.io/git-flow-cheatsheet/ 

A brief description of the git-flow workflow is here:

https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

and here is the original git-flow post: https://nvie.com/posts/a-successful-git-branching-model/

@chapmanb @naumenko-sa @hackdna @lpantano what do you all think about switching over to this method?

**Comment by lpantano** [comment]
I think is a good idea!

**Comment by hackdna** [comment]
Yes, I think it's a great way to manage a larger project like bcbio. This should also make upgrading more intuitive since `-u development` would actually pull from the develop branch instead of the master.

Btw, any more info on that bioconda bot?

**Comment by naumenko-sa** [comment]
+1. I agree this way it is is more organized. 
@Rory could you please clarify the interaction between bug fixing process and new development? Now when everything is happening in the master, people find and report bugs really fast, so they got fixed fast, because otherwise, the master is not working. Also, these bugfixes are one of the sources of new features. When we split bug fixing and dev, devs have to make sure they got the last bugfixes before merging, right? And they will need to spend more time testing, which was done by users before? S.

**Comment by roryk** [comment]
Hi Sergey,

With this method the `-u development` option will pull from the `develop` branch and `-u stable` will pull from `master`. The new `master` branch just has bugfixes. The new `develop` branch will be exactly what the `master` branch is now. It should still work, and should be bugfixed and also has new features in it.

I don't think it should add anything to our private testing, folks can still use `-u development` if they want to be on the cutting edge and folks running `-u stable` will still find bugs. This method should fix a bunch of issues that we get where the answer is 'just upgrade to development and it will fix it', since it will be already fixed on the stable branch.

**Comment by roryk** [comment]
@hackdna the new bioconda bot looks for releases on packages it knows about and automatically builds the package. So if we push regular hotfix type releases the bot will build the package for us, saving us a step. See https://github.com/bioconda/bioconda-recipes/pull/16410 as an example

**Comment by brainstorm** [comment]
As you mentioned, the [nvie model](https://nvie.com/posts/a-successful-git-branching-model/) is good, but I would prune all branches except `master` and `feature branches`. The (cognitive, newcomers) overhead of managing `development`, `hotfix`, etc .. is not worth it, in my experience.

`master` + `feature branches` is much faster (and you'll perhaps find that it does not break things that often), but as always YMMV.

/cc @vladsaveliev @ohofmann @pdiakumis

**Comment by roryk** [comment]
Great, I've been trying to develop using the git-flow model for the past few weeks, and I think I agree with Roman regarding keeping it the way it is with master + feature branches without the overhead of a separate develop branch for the time being. If we were doing a lot of feature development, or we had lots of folks developing, it might make more sense to move to that model. But right now it's just a few of us for the most part so the additional overhead isn't really useful, just aggravating. Thanks all for the comments!

**Comment by hackdna** [comment]
Just to illustrate the benefit of git-flow with a concrete example, this could have been applied as a hotfix to the stable version and thus not require the user to switch to the development version (which could contain other bugs): https://github.com/bcbio/bcbio-nextgen/issues/2795#issuecomment-517561579
